create mxnet quickstart doc (#655)
adap/flower: doc/source/quickstart_mxnet.rst

Quickstart (MXNet)

> ====================

Reviewer: ```suggestion
==================
```

In this tutorial we will learn how to train a Sequential Model on MNIST using Flower and MXNet. 

First of all, it is recommended to create a virtual environment and run everything within a `virtualenv <https://flower.dev/docs/recommended-env-setup.html>`_. 

Our example consists of one *server* and two *clients* all having the same model. 

*Clients* are responsible for generating individual model parameter updates for the model based on their local datasets. 

> These updates are then sent to the *server* which will aggregate them to produce a better model. Finally, the *server* sends this improved version of the model back to each *client*.

Reviewer: ```suggestion
These updates are then sent to the *server* which will aggregate them to produce an updated global model. Finally, the *server* sends this improved version of the model back to each *client*.
```

A complete cycle of parameters updates is called a *round*.

Now that we have a rough idea of what is going on, let's get started. We first need to install Flower. You can do this by running:

.. code-block:: shell

  $ pip install flwr

> Since we want to use MXNet, let's go ahead and install its library: 

Reviewer: ```suggestion
Since we want to use MXNet, let's go ahead and install it:
```

.. code-block:: shell

  $ pip install mxnet

Flower Client
-------------

> Now that we have all our dependencies installed, let's run a simple distributed training with two clients and one server. Our training procedure and network architecture are based on MXNet´s `Hand-written Digit Recognition <https://mxnet.apache.org/api/python/docs/tutorials/packages/gluon/image/mnist.html>`_. 

Reviewer: ```suggestion
Now that we have all our dependencies installed, let's run a simple distributed training with two clients and one server. Our training procedure and network architecture are based on MXNet´s `Hand-written Digit Recognition tutorial <https://mxnet.apache.org/api/python/docs/tutorials/packages/gluon/image/mnist.html>`_. 
```

In a file called :code:`client.py`, import Flower and MXNet related packages:

.. code-block:: python
      
    import flwr as fl

    import numpy as np

    import mxnet as mx
    from mxnet import nd
    from mxnet import gluon
    from mxnet.gluon import nn
    from mxnet import autograd as ag
    import mxnet.ndarray as F

> In addition, we define the device allocation in MXNet with:

Reviewer: ```suggestion
In addition, define the device allocation in MXNet with:
```

.. code-block:: python

    DEVICE = [mx.gpu() if mx.test_utils.list_gpus() else mx.cpu()]

We use MXNet to load MNIST, a popular image classification dataset of handrwritten digits for machine learning. The MXNet :code:`mx.test_utils.get_mnist()` downloads the training and test data. 

.. code-block:: python

    def load_data():
        print("Download Dataset")
        mnist = mx.test_utils.get_mnist()
        batch_size = 100
        train_data = mx.io.NDArrayIter(
            mnist["train_data"], mnist["train_label"], batch_size, shuffle=True
        )
        val_data = mx.io.NDArrayIter(mnist["test_data"], mnist["test_label"], batch_size)
        return train_data, val_data

> Define the training and loss with MXNet. The training of the dataset is done by looping over the dataset, measure the corresponding loss and optimize it. 

Reviewer: ```suggestion
Define the training and loss with MXNet. We train the model by looping over the dataset, measure the corresponding loss and optimize it. 
```

.. code-block:: python

    def train(net, train_data, epoch):
        trainer = gluon.Trainer(net.collect_params(), "sgd", {"learning_rate": 0.03})
        metric = mx.metric.Accuracy()
        softmax_cross_entropy_loss = gluon.loss.SoftmaxCrossEntropyLoss()
        for i in range(epoch):
            train_data.reset()
            for batch in train_data:
                data = gluon.utils.split_and_load(
                    batch.data[0], ctx_list=DEVICE, batch_axis=0
                )
                label = gluon.utils.split_and_load(
                    batch.label[0], ctx_list=DEVICE, batch_axis=0
                )
                outputs = []
                with ag.record():
                    for x, y in zip(data, label):
                        z = net(x)
                        loss = softmax_cross_entropy_loss(z, y)
                        loss.backward()
                        outputs.append(z)
                metric.update(label, outputs)
                trainer.step(batch.data[0].shape[0])
            name, acc = metric.get()
            metric.reset()
            print("training acc at epoch %d: %s=%f" % (i, name, acc))

> 

Reviewer: Remove one of the two empty lines here

> Define then the validation of the  machine learning network. We loop over the test set and measure the loss and accuracy of the test set. 

Reviewer: ```suggestion
Define then the validation of the  machine learning model. We loop over the test set and measure the loss and accuracy on the test set. 
```

            data = gluon.utils.split_and_load(batch.data[0], ctx_list=DEVICE, batch_axis=0)
            label = gluon.utils.split_and_load(
                batch.label[0], ctx_list=DEVICE, batch_axis=0
            )
            outputs = []
            for x in data:
                outputs.append(net(x))
                loss_metric.update(label, outputs)
                loss += loss_metric.get()[1]
            metric.update(label, outputs)
        print("validation acc: %s=%f" % metric.get())
        print("validation loss:", loss)
        accuracy = metric.get()[1]
        return loss, accuracy

After defining the training and testing of a MXNet machine learning model, we use the functions for the Flower clients.

The Flower clients will use a simple Sequential model:

.. code-block:: python

    def main():
        def model():
            net = nn.Sequential()
            net.add(nn.Dense(256, activation="relu"))
            net.add(nn.Dense(10))
            net.collect_params().initialize()
            return net

        train_data, val_data = load_data()

        model = model()
        init = nd.random.uniform(shape=(2, 784))
        model(init)

> 

Reviewer: Maybe add a comment about the model initialization here?

Author: done

> After loading the data set with :code:`load_data()` we define the Flower interface. 

Reviewer: ```suggestion
After loading the dataset with :code:`load_data()` we implement a Flower client. 
```
